### PR TITLE
Tracing: Tracing/Kernel dependency issue fix

### PR DESCRIPTION
**Pull request by tleksell-pe**
Added forward declarations of types declared
throughout kernel.h at the start of
kernel.h. With this change it is possible to
include tracing/tracing.h early in kernel.h
and use the tracing functionality in
kernel.h functions without compile errors,
which would result if tracing/tracing.h
was included at the end of kernel.h
(type dependencies).

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

### CODE REVIEW

**Comment by tleksell-pe** [comment]
This PR was spawned as part of a discussion in PR #40112 and aims to avoid some other trickery which was needed when tracing static inline functions in kernel.h (converting them into defines, see PR #40112). Neither I nor the Compliance check is very happy with the duplication of the typedefs so the question is if we should have the duplicates, move the existing ones, or not perform this change. Let me know what you think @dcpleung @andyross.

**Review by andyross — APPROVED** [review]
Looks great to me.  The only mild
criticism I have is that the commit
message is inexplicably filled at like
40 columns and looks like it was typed
on a Commodore 64.  Line length
preferences and hardware capabilities
vary, obviously, but I'm pretty sure
that in the modern world most people
reading this are going to have 80
column screens.

**Review by teburd — APPROVED** [review]
Looks good

**Review by dcpleung — APPROVED** [review]

**Comment by marc-hb** [comment]
Causes regression #40411 (with proprietary toolchain)

**Comment by marc-hb** [comment]
> Causes regression #40411 (with proprietary toolchain)

tl;dr: because of C11 `typedef` re-declarations.

C11 is being discussed in #30105
